src/code/sys_initial_check.c

```diff
@@ -1,17 +1,114 @@
+    Check_WriteRGBA16Pixel(buffer, x, y, value * GPACK_RGBA5551(16, 16, 16, 0) + GPACK_RGBA5551(12, 12, 12, 1));
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_initial_check/Check_ClearRGBA16.s")
+/**
+ * x and y are the coordinates of the bottom-left corner.
+ */
+void Check_DrawI4Texture(u16* buffer, s32 x, s32 y, s32 width, s32 height, u8* texture) {
+    s32 v;
+    s32 u;
+    u8 pixelPair;
+    u8* pixelPairPtr = texture;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_initial_check/Check_DrawExpansionPakErrorMessage.s")
+    // I4 textures are bitpacked 2 pixels per byte, so this writes a pair of pixels in each iteration using bitmasking.
+    for (v = 0; v < height; v++) {
+        for (u = 0; u < width; u += 2, pixelPairPtr++) {
+            pixelPair = *pixelPairPtr;
+            Check_WriteI4Pixel(buffer, x + u, y + v, pixelPair >> 4);
+            pixelPair = *pixelPairPtr;
+            Check_WriteI4Pixel(buffer, x + u + 1, y + v, pixelPair & 0xF);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_initial_check/Check_DrawRegionLockErrorMessage.s")
+void Check_ClearRGBA16(u16* buffer) {
+    u32 x;
+    u32 y;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_initial_check/Check_ExpansionPak.s")
+    for (y = 0; y < SCREEN_HEIGHT; y++) {
+        for (x = 0; x < SCREEN_WIDTH; x++) {
+            Check_WriteRGBA16Pixel(buffer, x, y, 1);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_initial_check/Check_RegionIsSupported.s")
+/**
+ * Draw error message textures directly to a screen buffer at the end of normal RDRAM
+ */
+void Check_DrawExpansionPakErrorMessage(void) {
+    DmaMgr_SendRequest0(CHECK_ERRMSG_STATIC_SEGMENT, SEGMENT_ROM_START(memerrmsg), SEGMENT_SIZE(memerrmsg));
+    Check_ClearRGBA16((u16*)FAULT_FB_ADDRESS);
+    Check_DrawI4Texture((u16*)FAULT_FB_ADDRESS, 96, 71, 128, 37, CHECK_ERRMSG_STATIC_SEGMENT);
+    Check_DrawI4Texture((u16*)FAULT_FB_ADDRESS, 96, 127, 128, 37,
+                        CHECK_ERRMSG_STATIC_SEGMENT + sizeof(gExpansionPakNotInstalledErrorTex));
+    osWritebackDCacheAll();
+    osViSwapBuffer((u16*)FAULT_FB_ADDRESS);
+    osViBlack(false);
+}
+
+/**
+ * Draw error message texture directly to a screen buffer at the end of normal RDRAM
+ */
+void Check_DrawRegionLockErrorMessage(void) {
+    DmaMgr_SendRequest0(CHECK_ERRMSG_STATIC_SEGMENT, SEGMENT_ROM_START(locerrmsg), SEGMENT_SIZE(locerrmsg));
+    Check_ClearRGBA16((u16*)FAULT_FB_ADDRESS);
+    Check_DrawI4Texture((u16*)FAULT_FB_ADDRESS, 56, 112, 208, 16, CHECK_ERRMSG_STATIC_SEGMENT);
+    osWritebackDCacheAll();
+    osViSwapBuffer((u16*)FAULT_FB_ADDRESS);
+    osViBlack(false);
+}
+
+/**
+ * If Expansion pak is not installed, display error message until console is turned off
+ */
+void Check_ExpansionPak(void) {
+    // Expansion pak installed
+    if (osMemSize >= 0x800000) {
+        return;
+    }
+
+    Check_DrawExpansionPakErrorMessage();
+    osDestroyThread(NULL);
+    while (true) {}
```

Comment:
For infinite loops like this one you can do this. 
```suggestion
    while (true);
```
I'm not saying it is better or not, up to you if you like it

Reply:
Wow C allows a lot of terrible syntax.